lib/AST/ASTImporter.cpp

```diff
@@ -1028,18 +1028,33 @@ void ASTNodeImporter::ImportDeclContext(DeclContext *FromDC, bool ForceImport,
     Importer.ImportContext(FromDC);
     return;
   }
-  llvm::SmallVector<Decl*, 8> Decls;
-  for (auto *From : FromDC->decls())
-    Decls.push_back(Importer.Import(From));
-  if (ToDC) {
-    // Restore the order.
-    for (auto *D : Decls) {
-      if (D && D->getLexicalDeclContext() == ToDC)
-        ToDC->removeDecl(D);
+  llvm::SmallVector<Decl*, 8> ImportedDecls;
+  for (Decl *From : FromDC->decls())
+    ImportedDecls.push_back(Importer.Import(From));
+
+  if (ToDC) { // Restore the order. O(n*n) in worst case, because the remove.
+
+    llvm::SmallDenseSet<Decl *, 2> MissingDecls;
+
+    // Frist, remove all declarations, which may be in wrong order.
+    for (Decl *ToD : ImportedDecls) {       // O(n)
+      if (ToD && ToDC->containsDecl(ToD)) { // containsDecl is O(1)
+        ToDC->removeDecl(ToD);              // O(n)
+      } else if (ToD) {
+        MissingDecls.insert(ToD);
+        FromDC->getParentASTContext().getDiagnostics().Report(
+            diag::warn_ast_importer_missing_decl_in_decl_context)
+            << ToD->getDeclKindName();
+      }
     }
-    for (auto *D : Decls) {
-      if (D && D->getLexicalDeclContext() == ToDC)
-        ToDC->addDeclInternal(D);
+    // Add the declarations, but this time in the correct order.
+    for (Decl *ToD : ImportedDecls) {
+      if (ToD) {
+        if (MissingDecls.count(ToD) ==
+            0) { // FIXME remove this if, when all Decls are properly imported
```

Comment:
I do not like the way the code and the comment is laid out.

Reply:
Clang format did it, but Ok, I will move the comment into line 1053, and hopefully it will look better.